### ISSUE TITLE
Do not auto-stop root dir detection at `.git`; make `.git` just another configurable root marker

### ISSUE THREAD

**Issue by LeoniePhiline** [helix-editor/helix#5852]
# Problem with nested repos / symlinked packages

My concrete problem was intelephense failing to resolve symbols in a php/composer/TYPO3 project.

Some composer packages which are shared between projects are installed into the tree. For development, they are symlinked from their original repo location.

Since they are symlinked, they do still contain their `.git` directory.

## Symlinked packages have a `.git`, and helix stops there

When editing a file of the symlinked package, helix still configures the LSP with the path of the symlinked package, because it contains a `.git` subfolder.

The problematic logic is implemented at https://github.com/helix-editor/helix/blob/master/helix-core/src/lib.rs#L74-L81 -- `.git` is hard-coded there. 

# Proposal

- Remove the special, hard-coded `.git` handling.
- Add `.git` to the default `roots` for all languages in `languages.toml`.

This way, the auto-discovery would still break at `.git` by default, but users can remove `.git` from their language `roots` configuration and have helix search up the file tree until the true root.

With the `break` removed (the entire `if ancestor.join(".git").exists()` branch removed), the existing algorithm would consume the entire iterator and return the outermost (closeset to filesystem root) directory matching any of the root markers.

`find_root()` from `helix-core` is used 

- for initializing an LSP
- for initializing a project file picker

IMHO, it makes the mistake of confusing "project" with "repository". The `root_markers` are a better way to determine the project root.

**Comment by pascalkuthe** [comment]
root markers are fundamentally distinct from `.git`.
`.git` indicates a workspace root and can not be expressed as another root marker for multiple reasons:

* The file picker shows files at `.git`
* Once file watching is implemented we will need to specify a root dir for watching (just watching every file in the FS is not feasible) this will aslo be the workspace root (`.git`)
* root dir search stops at `.git` and uses the last encountered `root marker`, this avoids constantly transversing the entire FS to the root

We should always stop looking for the root at the first `.git` directory as otherwise we need to always transverse the entire FS to find all root and somepeople have `.git` in their home dir (or just other forms of nested repos) and therefore transversing further than the first match would regess in those usecases.

However from your description I gathered that you don't even want to change the current workspace behavior. If I understand you correctly you have something like the following:

```
worksapace_dir
|
| - .git
| - subfolder
|    | - .git
|    | - foo.php
| bar.php
```

If I understand you right you want both `foo.php` and `bar.php` to use `workspace_dir` as your root but the `find_root` function is stopping at the `.git` of `subfolder` instead of `workspace_dir`. I do think that is a bug and can be improved. 
Instead of dynamically detecting the workspace root while searching for LSP roots we should just set the correct workspace directory (resolved from `cwd`) as a ceiling dir for `find_root`.

That is a reasonable request and fits well with the changes in #5748 which trys to adress similar issues. I will look into adding a fix to that PR

**Comment by LeoniePhiline** [comment]
> If I understand you right you want both `foo.php` and `bar.php` to use `workspace_dir` as your root but the `find_root` function is stopping at the `.git` of `subfolder` instead of `workspace_dir`.

This is exactly correct!

> I do think that is a bug and can be improved.

I'm very glad! :) This has been bugging me for a while.

> Instead of dynamically detecting the workspace root while searching for LSP roots we should just set the correct workspace directory (resolved from cwd) as a ceiling dir for find_root.

I would prefer a hypothetical solution in which my cwd can be `subfolder` and the project root at `workspace_dir` would still be correctly detected. With nested projects, paths to open files can become very long from the workspace root. (So long that the file picker cannot display the entire path.)

However, I can certainly live with having to open files from the workspace root. Either via `fd ... -X` or via helix' own file picker. Using cwd would definitely be a big improvement over stopping at `.git`.

> That is a reasonable request and fits well with the changes in https://github.com/helix-editor/helix/pull/5748 which trys to adress similar issues. I will look into adding a fix to that PR

Thank you! Can I help?

**Comment by pascalkuthe** [comment]
> I would prefer a hypothetical solution in which my cwd can be subfolder and the project root at workspace_dir would still be correctly detected

I don't think we can accomplish that. You need to consider that either the workspace root is `subfolder` in which case the fillepicker will show up there but the lsp will also stop at `subfolder` or the workspace root is `workspace_dir` in which case the LSP works but the picker  not quite.

I added a fix to #5748 so you can try that and see if it works for you. This actually lead me to find some other inconsistencies caused by handeling `root_dir` and `workspace_dir` in the same function so it's good you pointed this out.
